action/audit.go

```diff
@@ -3,76 +3,133 @@ package action
 import (
 	"fmt"
 	"os"
-	"strings"
+	"runtime"
 
 	"github.com/fatih/color"
 	"github.com/muesli/crunchy"
 	"github.com/muesli/goprogressbar"
 	"github.com/urfave/cli"
 )
 
+// AuditedSecret with its name, content a warning message and a pipeline error.
+type AuditedSecret struct {
+	name string
+
+	// the secret's content as a string. Needed for checking for duplicates.
+	content string
+
+	// message to the user about some flaw in the secret
+	message string
+
+	// real error that something in the pipeline went wrong
+	err error
+}
+
 // Audit validates passwords against common flaws
 func (s *Action) Audit(c *cli.Context) error {
 	t, err := s.Store.Tree()
 	if err != nil {
 		return err
 	}
+	list := t.List(0)
+
+	fmt.Printf("Checking %d secrets. This may take some time ...\n", len(list))
 
+	// Jobs are the secrets that still need auditing.
+	jobs := make(chan string)
+
+	// Secrets that have been audited.
+	checked := make(chan AuditedSecret)
+
+	// Spawn workers that run the auditing of all secrets concurrently.
 	validator := crunchy.NewValidator()
-	dupes := make(map[string][]string)
-	foundWeakPasswords := false
+	for worker := 1; worker <= runtime.NumCPU(); worker++ {
+		go s.audit(validator, jobs, checked)
+	}
 
-	pwList := t.List(0)
-	fmt.Printf("Checking %d secrets. This may take some time ...\n", len(pwList))
+	go func() {
+		for _, secret := range list {
+			jobs <- secret
+		}
+		close(jobs)
+	}()
+
+	duplicates := make(map[string][]string)
+	messages := make(map[string][]string)
 
 	bar := &goprogressbar.ProgressBar{
-		Text:    "Secrets checked",
-		Total:   int64(len(pwList)),
-		Current: 0,
-		Width:   120,
+		Total: int64(len(list)),
+		Width: 120,
 	}
-	for _, secret := range pwList {
-		bar.Current++
+
+	i := 0
+	for secret := range checked {
+		duplicates[secret.content] = append(duplicates[secret.content], secret.name)
+
+		if secret.message != "" {
+			messages[secret.message] = append(messages[secret.message], secret.name)
+		}
+
+		i++
+		bar.Current = int64(i)
 		bar.Text = fmt.Sprintf("%d of %d secrets checked", bar.Current, bar.Total)
 		bar.LazyPrint()
 
-		content, err := s.Store.GetFirstLine(secret)
-		if err != nil {
-			bar.Clear()
-			fmt.Println(color.RedString("Failed to retrieve secret '%s': %s", secret, err))
-			continue
+		if i == len(list) {
+			close(checked)
```

Comment:
True. I wasn't sure where and how to close the channel. But I probably can simple `break` there, true.
It's a bit tricky, because you can't really close the channel within the workers themselves.